### ISSUE TITLE
Evaluate commands (Shift-Enter/Return, Ctrl-Enter/Cmd-Return, menus) work not reliably

### ISSUE THREAD

**Issue by afischli** [supercollider/supercollider#1786]
The observations described subsequently are related to working through the tutorial (see also issue https://github.com/supercollider/supercollider/issues/1781). To repeat the bugs reported here try to execute code snippets from a Tutorial window, notably "02. First Steps" of Tutorial "Getting Started With SuperCollider" using a virgin SuperCollider.

**Shift-Enter/Shift-Return** works the most reliably under various states (see below), But even this shortcut not always. E.g. it does not work if the document from which one tries to execute some code was saved at some point in the past and does not have the proper extension '.scd', e.g. has extension '.txt'. 

**Ctrl-Enter/Cmd-Return** works 2nd most reliably but often fails, depending on the version of SC. With latest SC 3.7 it never does what it should with a code example from the Tutorial, e.g. "Hello world" or having selected two lines of code (2nd example in Tutorial), whereas Shift-Enter/Shift-Return works in the very same state.

Equivalent menu commands **"Language -> Evaluate Selection, Line or Region"** (Ctrl-Enter/Cmd-Return) and **"Language -> Evaluate Selection or Line"** (Shift-Enter/Shift-Return) also fail AFAIK always when the shortcut fails, but even fail more often and work the least reliably. There are cases, e.g. a code snippet in the Tutorial, where no evaluation/execution is possible via any of the evaluate menu commands, even not when the code snippet is selected, while some of their shortcuts work.

I tested this only on Mac systems. I found this buggy behavior in SC 3.7alpha1 (c651334), 3.6.6, 3.6.5 on 3 different Macs. I also found this behavior consistently using virgin accounts (no configurations active, no preference changed from Apple's default), using different keyboards (built in, Apple external, Logitech), under Mac OS X 10.6.8 (Snow Leopard), OS X 10.9.5 (Mavericks), and OS X 10.11.2 (El Capitan). All systems showed consistently the same bugs.

Moreover, **implicit autoselection behavior is state dependent**, i.e. differs in window Tutorial, Untitled window, or saved document Window. This is also confusing and it would be better to have a consistent behavior, notably regardless whether working from the help browser or the current document is unsaved (Untitled) or not or which extension the document has, as long as it is a document SC has been able to open and edit in etc. 

In summary: **All** the aforementioned **evaluate/execute commands (menus and their shortcuts) work properly as they should only if one executes them in a saved document with extension '.scd'**. **Otherwise some of the evaluation commands (shortcut or menu) fail, depending on SC's state**, i.e. focus, type of window (saved or untitled, extension dependent), currently a selection or not, and session, while none of those conditions should matter. Implicit autoselection behavior is also inconsistent and depends on state.

**Comment by jamshark70** [comment]
Thank you for testing on several systems.

I have no idea why nobody else is reporting similar bugs. Mac users of SuperCollider are not scarce. I have a student running SC on a Mac and he had no trouble at all running code.

This issue needs input from other Mac users.

**Comment by telephon** [comment]
sure, there are similar bugs: https://github.com/supercollider/supercollider/issues/1676
Also, the IDE (deliberately) switches off functionality when it encounters a non-scd / sc document. I'm not convinced that this is the best way, but it is currently the case.

**Comment by jamshark70** [comment]
@afischli Can you perhaps enumerate some specific scenarios? Bug reports generally go best with a "for dummies" set of steps to reproduce. That is, ironically despite the precision of your prose, after all this time I still simply can't work out the exact steps you're taking in your interface.

For instance, *With latest SC 3.7 it never does what it should with a code example from the Tutorial, e.g. "Hello world" or having selected two lines of code (2nd example in Tutorial), whereas Shift-Enter/Shift-Return works in the very same state.* Do you mean this?

1. Open the Getting Started tutorial, chapter 2.
2. Select the code example labeled "blah blah"?
3. Press cmd-return. Nothing happens.

That would fall under #1676.

Here is a sequence of steps that my student has tried, with (I believe) no problem, on a Mac. I *think* you're claiming this fails in your systems, but as it's buried in the middle of other cases, it's a bit fuzzy.

1. Launch the IDE.
2. Make a new code document.
3. Type `1+1` and hit cmd-return.

My student never complained that this was a problem. If this were universally a problem, every Mac user would have complained... but they don't. So there is some miscommunication going on.

**Comment by jamshark70** [comment]
> Also, the IDE (deliberately) switches off functionality when it encounters a non-scd / sc document.

That accounts for some of @afischli 's comments about "state." The suggestion is that all text files, regardless of extension, should trigger the same behavior in the ide. I don't know about that... I'm quite happy that emacs ignores this philosophy.

Perhaps what is missing is a visual indicator in the ide of editor "mode," then. That *would* be useful.

**Comment by afischli** [comment]
@jamshark70  You are getting at it. Sorry for not being clear about how to repeat the bugs. Yes one of the bugs I report here can be repeated exactly as you wrote:

1. Open the Getting Started tutorial, chapter 2.
2. Select the code example labeled "Hello World!"
3. Press cmd-return. Nothing happens.

Exactly. 

Sorry also for not having been aware of #1676, which actually describes the same thing for shortcut Cmd-Return. But I report much more than just a failing shortcut. Since the menu commands do also not work, while not being greyed out, which they should be if the state disables them. This would be a minimum of displaying the state change. But I do not understand what triggers exactly the state change and I doubt strongly this state change is actually intended. Moreover, I argue against it. I see little reason for that. The window in focus should be sufficient to tell SC which code to execute. Then it should do that reliably. Moreover, once having selected some code then all four commands (two menu, two shorcuts) should do exactly the same thing. All this is not the case in 3.7 and older versions of SC. Only 3.4.4 is to my knowledge behaving fine ihn this respect.

Then please don't argue that only a few Mac users have stumbled over this. You do not know how many were turned away from SC when they tried to use it a bit, i.e. they were following the Tutorial as I did, first steps and then encounter only things which don't work as described. If 'Hello World' does not work as described, why should you trust it. I also almost gave up on SC at that point.

Fortunately, in the meantime the likelihood of someone having the same problem should have gotten much smaller, since my suggestions to improve the tutorial, see #1787, has been merged to upstream (next release I hope). BTW, many thanks @telephon. Yet, the many bugs I report here are still there, inconsistencies. 

I have worked on theories about Human-Machine Interfaces in the past at lot and developed in 1985 the 'Dialog Machine' (DM) (and BTW also won awards for that work). The DM is a virtual machine, platform independent, as a layer between user and machine system. Based on that research I argue that every interactive program should have as few modal states as possible and if it is in a modal state, that needs to be made clear to the user. More importantly even, state changes need to be done consistently. Everything else is not helpful. SC fails exactly here. So, I argue 

(i) **SC IDE should have as few states as possible**. Now it seems to have several, where I argue one only would be fine. 
(ii) Then I argue **SC IDE does not handle its states**, whatever many there may be, **consistently**. As a consequence several **bugs emerge** from this.

Again an example easy to repeat (what I would call at least another bug than the one you can repeat in above 3 steps). 

1. Select some code snippet in a help browser window
2. press shortcut Shift-Return/Enter. The code will be executed. 
3. press shortcut Cmd-Retunr/Ctrl-Enter. The code will be NOT executed (above bug)
4. Now choose without touching anything (very same state) the menu command corresponding to Shift-Return/Enter. The code will be NOT executed (2nd bug)
5. Now choose without touching anything (very same state) the menu command corresponding to Shift-Return/Enter. The code will be NOT executed (perhaps a 3rd bug, at least its failure is consistent with what I called 1st bug)
6. Watch for the state of the two menu commands (since Shift-Return works, but Cmd-Return does not, the first corresponding menu command should be active (not greyed out), but the 2nd should be inactive (greyed out). This is not the case, both are active, i.e. not greyed out, yet do not work  accordingly to their state (perhaps a 4th bug)

etc. etc.

This makes all no sense and I call this an inconsistent handling of the very same state in terms of the human interface behavior. As a consequence, someone using a menu, while returning to SC after a while and having forgotten the shortcuts and therefore using menu commands for a while, will become completely confused when something as fundamental as executing some code in SC works for "magical" reasons sometimes and sometimes it doesn't. I am for sure one who hates that. IMHO way too much software is floating around, including commercial one, which shows such flaky behavior and one wastes all the time a lot of time with them for exactly that reason. Unreliable, inconsistent behavior!

Of course, I do not expect everyone to have similar views. But I hope you forgive me for explaining where I am coming from with this.

What to do? I am not sure. As I said, I tried in a pragmatic manner to make a quick fix by explaining in the tutorial better what is going on (#1787). Notably the very misleading exercise at the end of Step 2, in which I was invited to create a new document and save it and work with that is now avoided. Note, I followed religiously all what the tutorial suggested and towards the end I got so fed up with things not working at all as described so I opened issue https://github.com/supercollider/supercollider/issues/1781. This I believe will be avoided in the future with the next release of SC. Yet, IMHO the inconsistencies among the 4 aforementioned commands, IMHO clearly bugs, should to be fixed or somehow addressed better for sure.

**Comment by afischli** [comment]
Here some typical code with which I centrally ensure a consistent user interface in a DM program:

![setdmstate](https://cloud.githubusercontent.com/assets/11256493/12327965/8de915a2-bad9-11e5-9eeb-eb7c09974031.png)

I hope this is self-explanatory. So whenever some process in the program affects a state change, that routine 'SetDMState' needs to be called from wherever and whenever that may be. Then users can first properly see the state the program is in and also all menu commands and their associated shortcuts (in case of the DM that happens automatically) function consistently. 

In SC, at least in principal, there could be a similar object be instantiated. That object would have a similar method controlling the UI and any state change causing activity within SC would then send a message to that object, never doing anything with the UI itself, causing the user interface to be appropriately and always consistently updated. Is simpler to program, less likely to be buggy, and offers the user a pleasant reliable experience.

**Comment by jamshark70** [comment]
For clarity's sake, I would suggest moving all comments about **help browser** code execution problems to issue #1676, and reserve this issue for problems with execution within a code panel.

The code paths are quite different and I for one have been rather thoroughly confused by your description of the issues (which does not separate them). I thought you were talking about the code panel, which doesn't have most of the problems you describe. That's why I was pushing back.

AFAICS the only issue you raise with code panel execution is that you expected `.txt` to be OK for SC code documents, and it isn't.

Help browser execution indeed appears to be a mess. You're quite right to raise a bug report about those. I am not sure why my student didn't complain to me about that.

**Comment by afischli** [comment]
@jamshark70   Good idea. Yes, why not separating those in terms of finding pragmatic solutions. Indeed when it comes to "code panel execution" then I see there is only the issue that the intended behavior requires files to be saved and having the proper extension.

But please remember, again for clarity's sake, this issue again contains two different problems.

* One being saved but not having the extension .scd, e.g. having extension .txt (no execution possible, since any selection is overwritten)
* The other being not saved and being in state Untitled window/document where no extension whatsoever plays any role, since it has not yet been specified

Because if you have an Untitled window/document in which to execute code, then the behavior differs also from what you get when you save the document with extension .scd. I just did it again to make sure this is correct (using 3.7alpha1 under OS X 10.9.5 Mavericks). When the window is unsaved (Window/document Untitled) then Cmd-Return behaves like Shift-Return, which it shouldn't when the entire region is shown yellowed. That is it autoselects only the current line the cursor is on (ERROR: Command line parse failed -> nil). However, when the document is in saved state with proper extension .scd, then the behavior is fine and Cmd-Return autoselects the entire region and then executes it fine, while Shift-Return does autoselect only the current line and executes only that, which is then likely to fail. But the latter is of course irrelevant here.

Sorry for having only these bad "news". ;-)

**Comment by jamshark70** [comment]
> Yes, why not separating those in terms of finding pragmatic solutions.

The code panel is a Qt text editor widget (I forget its name -- I don't work intimately with the Qt objects in C++), while the help browser is a web page (displayed, iirc, by a QtWebKit) that uses JavaScript to run code examples. IMO the JavaScript was always a kludge and I wouldn't mind getting rid of it, but other developers felt it would make a better user experience to be able to interact with the help pages, even if the keyboard shortcuts didn't work as expected. Despite my argumentation here, actually I tend to agree that the broken keyboard shortcuts are worse for user experience: If we can't do code execution *properly* in help windows, then better not to do it at all. Anyway, that's for discussion in the other issue.

So, the code panel issues:

> One being saved but not having the extension .scd, e.g. having extension .txt (no execution possible, since any selection is overwritten).

This was explained. You may disagree, but the behavior is intentional. One reason is that you can edit .schelp documents in the ide, and you really don't want the editor's code lexer operating on English text. Automatic code selection depends on the lexer; hence, disabling the lexer also disables code execution. This could be clarified in the documentation, and the editor could be changed to show a symbol in the interface for non code tabs. But I don't expect the behavior to change. (Note, though, that code is colorized for sc and scd extensions, and not for txt. There is a visual cue, though probably too subtle.)

To summarize the other issue (for which your description is still too complex, too prose-y IMO). Please remember that we are all volunteers with other jobs and projects; having to read essays that mix technical description with editorial comment is likely to discourage participation. Note that the only developer who has spent any significant time on this (myself) is a Linux user who can't test or fix it!

So, if you agree with the following test case, I'll suggest closing this issue and opening a new one with *only* this case in it.

-  New code document ("Untitled").
- Type the following code block, exactly.
```
(
1+1;
1+2;
)
```
- Place the cursor on the `1+1` line.
- Press Cmd-Return.

Expected: It should run the whole block and post 3.

Actual: It runs only the single line and posts 2.

**Comment by jamshark70** [comment]
I asked Mac users to test: http://comments.gmane.org/gmane.comp.audio.supercollider.user/116425

"And compared their action to the menus and both work as advertised. So shift-return delivers 2 and cmd-return 3."

... which is what I was saying: The behavior reported in this issue is not happening for all Mac users. The proportion of Mac users for which it's a problem is unknown.

That will make it difficult to diagnose, I'm afraid... At least we can now say, there is some difference between @afischli 's systems and other Mac users' systems.

**Comment by miczac** [comment]
Hi, I should add: the test mentioned below I did on OS X 10.8.5.

**Comment by afischli** [comment]
Heureka, I found the culprit. More later, sorry, but I have to attend to other very urgent and important business. Will soon come back to this with a clean test case and I believe I can explain. The systems do not differ. ;-)

**Comment by afischli** [comment]
See #1781 for the promised test example.

**Comment by afischli** [comment]
Having closed #1781 I repeat from there here the steps to reproduce the bug in the parser:

```
(
1+1; // put the cursor into this line and press Cmd-Return => you get 3, press Shift-Return => you get 2 (All as expected)
1+2;
)

{ Saw.ar(LFNoise1.kr(1).range(100,[200,300]) }.play;

```
Above code causes no problems and all works as expected. However, if you reverse the ordering to this:

```
{ Saw.ar(LFNoise1.kr(1).range(100,[200,300]) }.play;

(
1+1; // put the cursor into this line and press Cmd-Return => you get 2 but should get 3, press Shift-Return => you get 2
1+2;
)

```
You no longer get what you expect, i.e. the parser fails to execute the correct statements within the region as defined by the parantheses. The reason is only the presence of a mistake before/above the region instead of after/below the region. The mistake is in the Saw.ar line, BTW a line taken from the Multichannel-Expansion guide (but now fixed cf. b54e617). But the parser should never get there in the first place, since that faulty line is outside the two parantheses and evaluation should stop at the left paranthesis and not encounter the line of code containing Saw.ar. 

I believe the parser should always evaluate correctly everything within a region as defined by parantheses regardless of what is outside of the region.

BTW, my guess is that this bug may also explain help browser code execution problems as described in issue #1676. But I have not studied that issue in details. It merely appears to me to be logical that if the parser is context dependent – as above reproducer demonstrates – that parsing may easily fail if a region is embedded in freely written text as is the case with help/tutorial texts.

**Comment by afischli** [comment]
@jamshark70 I believe this issue has nothing to do with a particular system. I have been able to reproduce the bug on several systems and using different versions of SC. The last test was done using SuperCollider 3.7.0-beta1 Built from branch 'HEAD' [9899190] (binary download) under OS X 10.9.5 (Mavericks).

**Comment by afischli** [comment]
Many thanks for all the ideas and help I got when trying to chase this bug. It was hard, I have to admit. However, it is never easy to chase a bug if some algorithm – here the parser – behaves context dependently.

To make testing by others perhaps more convenient and to express my gratitude a tiny little bit I attach a little zip archive ([#1786.zip](https://github.com/supercollider/supercollider/files/160288/1786.zip)) containing the test codes to reproduce the bug. ;-)

**Comment by afischli** [comment]
I close this issue, since I have opened as @telephon suggested a new follow up issue #1891.
